### PR TITLE
[FIXED JENKINS-38512] - "LinkageError: Failed to resolve null" when extensionpoint is an interface

### PR DESCRIPTION
**Pull request by ndeloof**
assuming we define `public interface Foo extends ExtensionPoint { ... }`
and later declare some `@Extension public static final foo WILL_FAIL = new Foo() { ... };`
jenkins fails to load with "LinkageError: Failed to resolve null"

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
🐝  and 👍 , but IMHO needs a JIRA issue to get it backported into 2.19.x

**Comment by ndeloof** [comment]
created JENKINS-38512 for this

**Comment by oleg-nenashev** [comment]
So squash and merge will link it properly now

**Comment by daniel-beck** [comment]
Is this actually a good idea? Interfaces as extension points seem to prevent evolving the API. Shouldn't those all be abstract classes?

**Comment by ndeloof** [comment]
@daniel-beck I don't think so. 
1. Some simple extension points (typically : Listeners) would be far simpler if expressed as interfaces
2. I expected Jenkins to move to java 8 at some point (I'm disappointed this didn't happened with 2.x), and abstract classes won't make much sense then.

**Comment by daniel-beck** [comment]
> Some simple extension points (typically : Listeners) would be far simpler if expressed as interfaces

[Here's a listener with more than half of its methods being added later (with `@since`)](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/slaves/ComputerListener.java).

> I [expect] Jenkins to move to java 8 at some point…, and abstract classes won't make much sense then.

Then let's implement this change then. Right now, Jenkins is Java 7, and we shouldn't introduce something like this that has obvious problems in that environment.

**Comment by ndeloof** [comment]
@daniel-beck All those methods could have been added to specific listeners, as Functional Interfaces, if we had consideration for future move to Java 8.
Anyway, the question here is not to know if this is a bad practice, or if we want Jenkins to badly fail to start when someone adopt it

**Comment by daniel-beck** [comment]
>  if we want Jenkins to badly fail to start when someone adopt it

Given its significant drawbacks, I'd rather see Jenkins break immediately so that nobody can even do this successfully, than have API evolution problems a year down the road when there's no way to go back, and, without increasing the JRE requirement, no way to go forward.

That said, I'd like to see further opinions here. @jenkinsci/code-reviewers ?

**Comment by ndeloof** [comment]
@daniel-beck What makes you think using interfaces means no way to go forward ? We use to provide compatibility code inside abstract method, using interfaces would just move this constraint to method caller, typically considering Listeners, we would need fireXX to consider all possible listeners to match an event, and invoke adequate interfaces' methods including deprecated ones.

**Comment by oleg-nenashev** [comment]
Well, I'm not a big fan of Interfaces as extension points. Just because adding extensibility will likely require creating more and more classes. But maybe we should allow plugin maintainers to shoot in their leg if they feel strongly. But further maintainers may be not so happy if the plugin gets handed over...

The PR itself is good. Maybe such high-level topic needs a wider discussion in the ML since CCing @jenkinsci/code-reviewers is generally useless

**Comment by oleg-nenashev** [comment]
Added it to the Governance meeting agenda

**Review by jglick — APPROVED** [review]
This needs a test, but otherwise :+1: on it.

**Review by jglick — CHANGES_REQUESTED** [review]
Sorry, meant to “request changes”.

**Comment by jglick** [comment]
> I'd rather see Jenkins break immediately so that nobody can even do this successfully

AFAIK that ship has sailed. You can use interfaces as extension points today (and I have done so at least in prototype code)—just not, apparently, with extensions implemented as fields initialized with anonymous inner classes, a relatively exotic form. So from my PoV this is simply a bug that should be fixed.

**Comment by daniel-beck** [comment]
> You can use interfaces as extension points today

Oh well.

If that's the case, seems fine.

**Comment by batmat** [comment]
Maybe this is one reason to push Java 8 as baseline going forward, so that we can start using default methods. This is not a solution for everything, but could most probably help in that area.

**Comment by jglick** [comment]
> so that we can start using default methods. This is not a solution for everything

Indeed, default methods allow interfaces to “perform” as “well” as (pure) abstract classes in just one, albeit one common, situation: that you wanted to convert

```java
interface Provider extends ExtensionPoint {
    Stuff provide();
}
```

to

```java
interface Provider extends ExtensionPoint {
    default boolean threadSafe() {
        return false;
    }
    Stuff provide();
}
```

But this happens to a situation that you can already easily handle; in fact all sorts of API refactorings are easy enough when it comes to extension points themselves:

```java
@Deprecated
interface Provider extends ExtensionPoint {
    Stuff provide();
}
interface Provider2 extends ExtensionPoint {
    boolean threadSafe();
    Optional<Stuff> provide(Thing context);
}
```

Because extension points are generally not referred to in the signatures of other API methods, it is easy to migrate them: old implementations will get deprecation warnings; the caller of `ExtensionList.lookup` can simply iterate the old implementations and deal with them however it sees fit.

In other words, the stronger argument for avoiding `interface` in favor of `abstract class` comes only when you are defining a type that is going to be passed around in complex ways, making it impractical to simply replace it (since you would need to transitively deprecate and replace a much larger API to match).

And anyway if you wanted to block people from _defining_ extension points as interfaces, the right approach would have been an annotation processor that makes the API uncompilable—rather than relying on an undocumented runtime exception to make _implementations_ not work (and only under relatively obscure conditions).

**Comment by ndeloof** [comment]
my intent was a bit distinct. Let's consider for sample `RunListener`. This abstract class has been designed so we can add new events but keep compatibility with empty implementation. As a result a component to use this API has to create a dedicated class and override method. As a result one can't have same component to extend another API.

Let's now say we model RunListener as functional interfaces:
```
@FunctionalInterface
public interface RunStartedListener {
    void onStarted(R r, TaskListener listener);
}
```

Any component can now implement this interface, and be identified by jenkins-core as a listener for this _single_ event. If they need to listen to some more event, just implement the other interfaces. We can extend the API with additional interfaces, no need for a `RunListener2` approach. Just the caller has to lookup jenkins component for a specific interface. This also open the door to lambdas.

Also, if we have API method that need to _receive_ such a component as interface, but require more that just one, we can use:
```
   public void someAPIMethod(<? extends SomeInterface & SomeOtherInterface> foo) { ... }
```

We could for sample rely on this to avoid classes like AbstractProject with tons of unrelated implementation details, and only require some `<? extends Actionable & WithSCM>` in SCM's `compareRemoteRevision`

**Comment by stephenc** [comment]
:bee: and 👍  on the idea... :bug: on the lack of test case

**Comment by oleg-nenashev** [comment]
@ndeloof Any plans to address the bug from @stephenc ?

**Comment by oleg-nenashev** [comment]
@ndeloof ping

**Comment by ndeloof** [comment]
nop
